### ISSUE TITLE
TODO: add new package shadow

### ISSUE THREAD

**Issue by odoublewen** [spack/spack#12297]
Need new package for `shadow` project

* package `singularity@3.3.0:` depends on command `newuidmap` (and `newguimap`) for full functionality (fakeroot support).
* `newuidmap` is provided by upstream project https://github.com/shadow-maint/shadow
* building `shadow` in turn appears to require `semange.h`

@vsoch should this package be called `shadow-uidmap` or just `shadow`?

See #12295 for motivation

**Comment by vsoch** [comment]
I think shadow-uitmap, just "shadow" is really general, and we should be specific!

**Comment by vsoch** [comment]
hey @odoublewen I'm giving this a try, and I'm hitting this error:

```
==> Error: ProcessError: Command exited with status 1:
    './autogen.sh'

2 errors found in build log:
     55    libtoolize: Consider adding 'AC_CONFIG_MACRO_DIRS([m4])' to configure.ac,
     56    libtoolize: and rerunning libtoolize and aclocal.
     57    libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
     58    configure.ac:626: warning: macro 'AM_GNU_GETTEXT_VERSION' not found in library
     59    configure.ac:627: warning: macro 'AM_GNU_GETTEXT' not found in library
     60    autoreconf: running: /home/vanessa/Desktop/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/autoconf-2.69-cqde5ueevm
           ks5n4awekqmnlxirjfj2km/bin/autoconf --force
  >> 61    configure.ac:626: error: possibly undefined macro: AM_GNU_GETTEXT_VERSION
     62          If this token and others are legitimate, please use m4_pattern_allow.
     63          See the Autoconf documentation.
  >> 64    configure.ac:627: error: possibly undefined macro: AM_GNU_GETTEXT
     65    autoreconf: /home/vanessa/Desktop/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.4.0/autoconf-2.69-cqde5ueevmks5n4awek
           qmnlxirjfj2km/bin/autoconf failed with exit status: 1

See build log for details:
  /tmp/vanessa/spack-stage/shadow-uidmap-4.6-ozsxmu2ctqantwyej5cdiauvvyrlddwc/spack-build-out.txt
```
Any ideas for what to try? The various google searches / solutions offered didn't seem to work, but maybe I missed the right one! :)

**Comment by odoublewen** [comment]
Can I see what you have so far?  Do you have spack forked somewhere?  Not sure I will be much use, but I'm usually willing to hit my head against a wall a little -- occasionally I find a loose brick or two :)

**Comment by vsoch** [comment]
Yes!

```python
# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
# Spack Project Developers. See the top-level COPYRIGHT file for details.
#
# SPDX-License-Identifier: (Apache-2.0 OR MIT)

from spack import *


class ShadowUidmap(AutotoolsPackage):
    """Tools to help unprivileged users create uid and gid mappings in 
       user namespaces."""

    homepage = "https://github.com/shadow-maint/shadow"
    url      = "https://github.com/shadow-maint/shadow/archive/4.6.tar.gz"
    git      = "https://github.com/shadow-maint/shadow.git"

    version('4.6', sha256='045c3992f0e0f62e893ddd8bc20a3c6a4a7ed0bf59c18c7da1fae13193e3f2d7')
    version('devel', branch='master')
    version('v4.1.2.1', branch='v4.1.2.1')
    version('v4.1.2.2', branch='v4.1.2.2')
    version('v4.1.4.3', branch='v4.1.4.3')

    depends_on('m4',       type='build')
    depends_on('autoconf', type='build')
    depends_on('automake', type='build')
    depends_on('libtool',  type='build')
    depends_on('gettext')

    def autoreconf(self, spec, prefix):
        autogen = Executable('./autogen.sh')
        autogen()
        #automake = which('automake')
        #automake('-a') # force missing and add missing
        #autoreconf = which('autoreconf')
        #autoreconf('-ifv')

    def install(self, spec, prefix):
        make('install', parallel=False)
```

The commented out bits are what I've been playing around with - I had thought that adding flags to those commands could help (but haven't so far).

**Comment by vsoch** [comment]
And here it is packaged with a branch: https://github.com/researchapps/spack/blob/add/shadow-uidmap/var/spack/repos/builtin/packages/shadow-uidmap/package.py

Sorry it took me a bit - I had cloned with depth 1 and had to redo it!

**Comment by odoublewen** [comment]
OK, so `configure` is failing due to missing `AM_GNU_GETTEXT` macro.

The `autoreconf` command is giving an early warning of the same thing.

```console
$ cd /tmp/odoublewen/spack-stage/shadow-uidmap-4.6-yws5gqsgz73gynhresggxjynrhtt44dn
$ . spack-build-env.txt
$ cd spack-src/
$ ./autogen.sh
autoreconf: Entering directory `.'
autoreconf: running: autopoint --force
autoreconf: running: aclocal --force 
configure.ac:626: warning: macro 'AM_GNU_GETTEXT_VERSION' not found in library
configure.ac:627: warning: macro 'AM_GNU_GETTEXT' not found in library
autoreconf: configure.ac: tracing
autoreconf: configure.ac: AM_GNU_GETTEXT_VERSION is used, but not AM_GNU_GETTEXT
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: Consider adding 'AC_CONFIG_MACRO_DIRS([m4])' to configure.ac,
libtoolize: and rerunning libtoolize and aclocal.
libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
configure.ac:626: warning: macro 'AM_GNU_GETTEXT_VERSION' not found in library
configure.ac:627: warning: macro 'AM_GNU_GETTEXT' not found in library
autoreconf: running: /home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/autoconf-2.69-gvwf2dujnyrmomflwe5oq3jhidfdf7c7/bin/autoconf --force
configure.ac:626: error: possibly undefined macro: AM_GNU_GETTEXT_VERSION
      If this token and others are legitimate, please use m4_pattern_allow.
      See the Autoconf documentation.
configure.ac:627: error: possibly undefined macro: AM_GNU_GETTEXT
autoreconf: /home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/autoconf-2.69-gvwf2dujnyrmomflwe5oq3jhidfdf7c7/bin/autoconf failed with exit status: 1
```

I see you have gettext listed as a dependency.  In my build environment, it does get added to the `CMAKE_PREFIX_PATH`, but from what I can tell online, the macro needs to be defined in one aclocal's `.m4` files.  I don't see any mention of this macro in those files.

```console
$ aclocal --print-ac-dir
/home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/automake-1.16.1-h5z4w6lujja35hm3sgfqzlobmbn5bosq/share/aclocal

$ ls /home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/automake-1.16.1-h5z4w6lujja35hm3sgfqzlobmbn5bosq/share/aclocal*
/home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/automake-1.16.1-h5z4w6lujja35hm3sgfqzlobmbn5bosq/share/aclocal:
README

/home/odoublewen/spacktree/spack/opt/spack/linux-centos6-x86_64/gcc-8.2.0/automake-1.16.1-h5z4w6lujja35hm3sgfqzlobmbn5bosq/share/aclocal-1.16:
amversion.m4  cond.m4          gcj.m4         lex.m4         mkdirp.m4       runlog.m4    tar.m4
ar-lib.m4     depend.m4        init.m4        lispdir.m4     obsolete.m4     sanity.m4    upc.m4
as.m4         depout.m4        install-sh.m4  maintainer.m4  options.m4      silent.m4    vala.m4
auxdir.m4     dmalloc.m4       internal       make.m4        prog-cc-c-o.m4  strip.m4
cond-if.m4    extra-recurs.m4  lead-dot.m4    missing.m4     python.m4       substnot.m4
```

@hartzell Any ideas?  :)

**Comment by vsoch** [comment]
Yeah, most of the instructions online came down to editing the configuration files themselves, which didn't seem like the right/best solution.

**Comment by odoublewen** [comment]
Yes.  The gettext .m4 files are in their opt directory:

```
find opt/spack/linux-centos6-x86_64/gcc-8.2.0/gettext-0.19.8.1-vtt4qdhwuaedpphtikxd44tzoc4jnfgu/ -name 'gettext.m4'
opt/spack/linux-centos6-x86_64/gcc-8.2.0/gettext-0.19.8.1-vtt4qdhwuaedpphtikxd44tzoc4jnfgu/share/aclocal/gettext.m4
```

...and if we add the `gettext-0.19.8.1-vtt4qdhwuaedpphtikxd44tzoc4jnfgu/share/aclocal` directory to the autoreconf command:

```
autoreconf -v -f -I path/to/gettext-0.19.8.1-xyz/share/aclocal --install
```

Then we get a lot further into the build.  Failing now due to missing semanage headers.  (selinux headers are okay).

```
checking for selinux/selinux.h... yes
checking semanage/semanage.h usability... no
checking semanage/semanage.h presence... no
checking for semanage/semanage.h... no
configure: error: semanage/semanage.h is missing
```

**Comment by vsoch** [comment]
Ah, let me see what I can do with that!

**Comment by vsoch** [comment]
I got it working, but for some reason I didn't hit that semanage.h error... is it because I have ubuntu?

**Comment by hartzell** [comment]
This works for me on Owen's system (CentOS 7):

```python
  # Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
 # Spack Project Developers. See the top-level COPYRIGHT file for details.
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)

 from spack import *


 class Shadow(AutotoolsPackage):
     """Tools to help unprivileged users create uid and gid mappings in
        user namespaces."""

     homepage = "https://github.com/shadow-maint/shadow"
     url      = "https://github.com/shadow-maint/shadow/releases/download/4.7/shadow-4.7.tar.gz"

     version('4.7', sha256='5135b0ca2a361a218fab59e63d9c1720d2a8fc1faa520c819a654b638017286f')
     version('4.6', sha256='4668f99bd087399c4a586084dc3b046b75f560720d83e92fd23bf7a89dda4d31')
```

Associated commentary:

- The package should probably be named "shadow", because:
  1. That's what `spack create` names the package when you hand it the download URL; and
  2. That's what the upstream package is named.
  2. I'm pretty sure that that's what @adamjstewart would tell me to do...
- When dealing with things distributed on GitHub, you should prefer `download` URLs to `archive` URLS.  
  The download URL's contain the "release artifacts" that the developers uploaded to go along with the release, they're generally the half-baked pizza that they expect end users to compile at home.  
  On the other hand, the contents of the archive URLs are the straight contents of the git repo at that point, heaven only knows what magic the developers undertake to make a release from it (ok, they may have docs, but....)
- Is there a pressing need for a `devel` release to build the latest greatest from the tip of master?

I can PR this if you'd like.

**Comment by vsoch** [comment]
I’ve been working really hard on this and I already opened a PR, would you mind pushing to my branch so we both are contributors? Mine worked too, but as you point out, it’s flawed :(

**Comment by vsoch** [comment]
huh, I had originally chosen the download URLs but changed to package when I saw it used in another package. I wonder if that one is messed up too.

Anyway, I'm happy to change my PR, or if you want credit, you can push to my branch. I'd like to be involved, if that's ok with you...

**Comment by vsoch** [comment]
okay I fixed it, sorry for being such an idiot.

https://github.com/spack/spack/pull/12750

**Comment by hartzell** [comment]
Go for it!

**Comment by hartzell** [comment]
> huh, I had originally chosen the download URLs but changed to package when I saw it used in another package. I wonder if that one is messed up too.

There are reasons to do it both ways, depending.  In general using the download URLs is *lazier*, which as [Larry says](http://threevirtues.com/) is more *virtuous*.  But the devs might not have uploaded half-baked pizzas (yeah, it's lunch time, food analogies abound), you might need to rebuild from scratch and the devs didn't include all of the raw materials in the `download` tarballs, or perhaps  you need to build older versions that don't *have* `download` tarballs (but, do you really need them?).

One of the big downsides of the `archive` tarballs is that they're produced on demand (cached, I'm sure).  Every once in a while GitHub decides to change something about how they produce the tarballs and the end result is that the checksums no longer match.  You'll see big swaths of commits where someone updates many, many package signatures....

**Comment by odoublewen** [comment]
> This works for me on Owen's system (CentOS 7)

Actually, I was sandboxing on an old centos7 system.  I think that explains why I didn't have the needed SE manage goodies.

**Comment by adamjstewart** [comment]
Assuming this is closed by #12750, reopen if not.
